--- fancontrol-new	2009-06-16 01:48:03.000000000 +0200
+++ /usr/sbin/fancontrol	2009-06-16 02:56:31.000000000 +0200
@@ -3,7 +3,7 @@
 # Simple script implementing a temperature dependent fan speed control
 # Supported Linux kernel versions: 2.6.5 and later
 #
-# Version 0.68
+# Version 0.7
 #
 # Usage: fancontrol [CONFIGFILE]
 #
@@ -78,10 +78,15 @@
 		exit 1
 	fi
 
+	steptime="0.25" # use an interval thats fast enough to not be heard during ramp-up/down 
+	stepcount=$(echo "scale=0; $INTERVAL / $steptime;" | bc ) # calculate needed steps, but discard decimal tail
+
 	# write settings to arrays for easier use and print them
 	echo
 	echo "Common settings:"
 	echo "  INTERVAL=$INTERVAL"
+	echo "  STEPTIME=$steptime"
+	echo "  STEPCOUNT=$stepcount"
 
 	let fcvcount=0
 	for fcv in $FCTEMPS
@@ -326,15 +331,47 @@
 			wait $!
 		  fi
 		fi
-		echo $pwmval > $pwmo # write new value to pwm output
-		if [ $? -ne 0 ]
-		then
-			echo "Error writing PWM value to $DIR/$pwmo"
-			restorefans 1
-		fi
-		if [ "$DEBUG" != "" ]
-		then
-			echo "new pwmval=$pwmval"
+		
+		if [ $(($pwmpval - $pwmval)) -gt 1 -o $(($pwmpval - $pwmval)) -lt -1 ]; then # only bother when theres a significant change to conserve cpu time
+
+			pwmstep=$(echo "scale=2; -( $pwmpval - $pwmval ) / $stepcount;" | bc) # pwmstep = pwm change in one step
+
+			if [ "$DEBUG" != ""  ]
+			then
+				echo "target pwmval=$pwmval"
+				echo "pwmstep=$pwmstep"
+			fi
+
+			for i in $(seq 1 $stepcount) # have $i count from 1 to $stepcount
+			do
+				pwmval=$(echo "scale=0; $pwmpval + ( $i * $pwmstep ) / 1;" | bc) # increase pwmpval by $pwmstep
+
+				if [ "$DEBUG" != "" ]
+				then
+					echo "new pwmval=$pwmval"
+				fi
+
+				echo $pwmval > $pwmo # write new value to pwm output 
+				# Sleep while still handling signals
+				sleep $steptime &
+				wait $!
+			done
+
+
+			if [ $? -ne 0 ]
+			then
+				echo "Error writing PWM value to $DIR/$pwmo"
+				restorefans 1
+			fi
+		else
+			if [ "$DEBUG" != "" ]
+			then
+				echo "target pwmval=$pwmval=~pwmpval"
+			fi
+			# Sleep while still handling signals
+			sleep $INTERVAL &
+			wait $!
+
 		fi
 		let fcvcount=$fcvcount+1
 	done
@@ -361,6 +398,6 @@
 do
 	UpdateFanSpeeds
 	# Sleep while still handling signals
-	sleep $INTERVAL &
+	sleep 0.05 &
 	wait $!
 done
\ No newline at end of file
